aspnetcore/client-side/dotnet-js-interop.md

```diff
@@ -0,0 +1,830 @@
+    // A long/Int64 can be mapped as either a Number or BigInt.
+    // Passing a long value to the above method will generate an error "ToJS for System.Int64 is not implemented." at runtime.
+    // If the parameter declaration `Method(JSMarshalAs<JSType.Any>] long value)` is used, then a compile time error is generated: "Type long is not supported by source-generated JavaScript interop...."
+    // Instead, map the long parameter explicitly to either a JSType.Number or JSType.BigInt.
+    // Note there could potentially be runtime overflow errors in JS if the C# value is too large.
+    [JSImport("PrimitivesShim.LogValueAndType", "PrimitivesShim")]
+    public static partial void LogValueAndTypeForNumber([JSMarshalAs<JSType.Number>] long value);
+
+    [JSImport("PrimitivesShim.LogValueAndType", "PrimitivesShim")]
+    public static partial void LogValueAndTypeForBigInt([JSMarshalAs<JSType.BigInt>] long value);
+}
+
+public static class PrimitivesUsage
+{
+    public static async Task Run()
+    {
+        // Ensure JS ES6 module loaded
+        await JSHost.ImportAsync("PrimitivesShim", "/PrimitivesShim.js");
+
+        // Call a proxy to a static JS method, console.log("")
+        PrimitivesProxy.ConsoleLog("Printed from JSImport of console.log()");
+
+        // Basic examples of JS interop with an integer:       
+        PrimitivesProxy.IncrementCounter();
+        int counterValue = PrimitivesProxy.GetCounter();
+        PrimitivesProxy.LogInt(counterValue);
+        PrimitivesProxy.LogString("I'm a string from .NET in your browser!");
+
+        // Mapping some other .NET types to JS primitives:        
+        PrimitivesProxy.LogValueAndType(true);
+        PrimitivesProxy.LogValueAndType(0x3A);// byte literal
+        PrimitivesProxy.LogValueAndType('C');
+        PrimitivesProxy.LogValueAndType((Int16)12);
+        // Note: JavaScript Number has a lower max value and can generate overflow errors
+        PrimitivesProxy.LogValueAndTypeForNumber(9007199254740990L);// Int64/Long 
+        PrimitivesProxy.LogValueAndTypeForBigInt(1234567890123456789L);// Int64/Long, JS BigInt supports larger numbers
+        PrimitivesProxy.LogValueAndType(3.14f);// single floating point literal
+        PrimitivesProxy.LogValueAndType(3.14d);// double floating point literal
+        PrimitivesProxy.LogValueAndType("A string");
+    }
+}
+// The example displays the following output in the browser's debug console:
+//       Printed from JSImport of console.log()
+//       1
+//       I'm a string from .NET in your browser!
+//       boolean true
+//       number 58
+//       number 67
+//       number 12
+//       number 9007199254740990
+//       bigint 1234567890123456789n
+//       number 3.140000104904175
+//       number 3.14
+//       string A string
+```
+
+## JS Date
+
+Demonstrates JSImport of methods which have a JS Date object as its return or parameter.
```

Comment:
I also wondered if I should mention that even though in JS this is an object type, it is marshalled by-value rather than by-reference?  Wasn't sure on accurate wording for that though.

Reply:
in JS the it's object around `double`/`Number` anyway. But yeah, it's marshaled as value type.

Reply:
"marshaled by value" is ok

Reply:
Updated Date example intro.  Probably too wordy but I was trying to make it explicit enough that someone not familiar with nuances would have an easier time grasping:

> Demonstrates JSImport of methods which have a JS Date object as its return or parameter.  Dates are marshalled across interop by-value, meaning they are copied in much the same way as JS primitives.
> 
> Be aware that JS Date objects do not contain timezone information and are internally represented in UTC.  To ensure a reliable conversion when marshalling to JavaScript, a .NET DateTime should be initialized with DateTimeKind.Utc or DateTimeKind.Local, avoiding DateTimeKind.Unspecified.  Consider a scenario where the local environment's timezone is EST. If you pass the DateTime `new DateTime(1968, 12, 21, 7, 51, 0, DateTimeKind.Local)` to JavaScript and it were later returned back to .NET, then it will have DateTimeKind.Utc with the time adjusted to 12:51 PM UTC.  Although the original DateTimeKind was changed, the time is adjusted correctly.

https://github.com/SerratedSharp/AspNetCore.Docs/pull/1/commits/88ee0abdf34825a3f11c0cc3f4e666aa720831f8